File: core/core-sync/src/main/java/ch/difty/scipamato/core/sync/launcher/RefDataSyncJobLauncher.java
package ch.difty.scipamato.core.sync.launcher;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;

import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.batch.core.*;
import org.springframework.batch.core.launch.JobLauncher;
import org.springframework.batch.core.repository.JobExecutionAlreadyRunningException;
import org.springframework.batch.core.repository.JobInstanceAlreadyCompleteException;
import org.springframework.batch.core.repository.JobRestartException;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

/**
 * Launcher for the reference dat sync job. Comprises of the following
 * synchronization jobs:
 * <ol>
 * <li>languages</li>
 * <li>NewStudyPage links</li>
 * <li>code_classes</li>
 * <li>codes</li>
 * <li>papers</li>
 * <li>newsletters</li>
 * <li>newsletterTopics</li>
 * <li>newStudyTopics</li>
 * <li>newStudies</li>
 * <li>keywords</li>
 * </ol>
 *
 * @author u.joss
 */
@Slf4j
@Component
@Profile("!wickettest")
public class RefDataSyncJobLauncher implements SyncJobLauncher {

    private final Job         syncLanguageJob;
    private final Job         syncNewStudyPageLinkJob;
    private final Job         syncCodeClassJob;
    private final Job         syncCodeJob;
    private final Job         syncPaperJob;
    private final Job         syncNewsletterJob;
    private final Job         syncNewsletterTopicJob;
    private final Job         syncNewStudyJob;
    private final Job         syncNewStudyTopicJob;
    private final Job         syncKeywordJob;
    private final JobLauncher jobLauncher;

    private final Warner warner;

    public RefDataSyncJobLauncher(@NotNull final JobLauncher jobLauncher,
        @Qualifier("syncLanguageJob") @NotNull final Job syncLanguageJob,
        @Qualifier("syncNewStudyPageLinkJob") @NotNull final Job syncNewStudyPageLinkJob,
        @Qualifier("syncCodeClassJob") @NotNull final Job syncCodeClassJob,
        @Qualifier("syncCodeJob") @NotNull final Job syncCodeJob,
        @Qualifier("syncPaperJob") @NotNull final Job syncPaperJob,
        @Qualifier("syncNewsletterJob") @NotNull final Job syncNewsletterJob,
        @Qualifier("syncNewsletterTopicJob") @NotNull final Job syncNewsletterTopicJob,
        @Qualifier("syncNewStudyJob") @NotNull final Job syncNewStudyJob,
        @Qualifier("syncNewStudyTopicJob") @NotNull final Job syncNewStudyTopicJob,
        @Qualifier("syncKeywordJob") @NotNull final Job syncKeywordJob, @NotNull final Warner warner) {
        this.jobLauncher = jobLauncher;
        this.syncNewStudyPageLinkJob = syncNewStudyPageLinkJob;
        this.syncLanguageJob = syncLanguageJob;
        this.syncCodeClassJob = syncCodeClassJob;
        this.syncCodeJob = syncCodeJob;
        this.syncPaperJob = syncPaperJob;
        this.syncNewsletterJob = syncNewsletterJob;
        this.syncNewsletterTopicJob = syncNewsletterTopicJob;
        this.syncNewStudyJob = syncNewStudyJob;
        this.syncNewStudyTopicJob = syncNewStudyTopicJob;
        this.syncKeywordJob = syncKeywordJob;
        this.warner = warner;
    }

    @NotNull
    @Override
    public SyncJobResult launch() {
        log.info("Starting synchronization job from scipamato-core to scipamato-public...");
        final SyncJobResult result = new SyncJobResult();
        final JobParameters jobParameters = getJobParameters();
        try {
            warnAboutUnsynchronizedEntities(result);

            runSingleJob("languages", syncLanguageJob, result, jobParameters);
            runSingleJob("newStudyPage_links", syncNewStudyPageLinkJob, result, jobParameters);
            runSingleJob("code_classes", syncCodeClassJob, result, jobParameters);
            runSingleJob("codes", syncCodeJob, result, jobParameters);
            runSingleJob("papers", syncPaperJob, result, jobParameters);
            runSingleJob("newsletters", syncNewsletterJob, result, jobParameters);
            runSingleJob("newsletterTopics", syncNewsletterTopicJob, result, jobParameters);
            runSingleJob("newStudyTopics", syncNewStudyTopicJob, result, jobParameters);
            runSingleJob("newStudies", syncNewStudyJob, result, jobParameters);
            runSingleJob("keywords", syncKeywordJob, result, jobParameters);

            if (result.isSuccessful())
                log.info("Job finished successfully.");
            else
                log.warn("Job finished with issues.");
        } catch (final Exception ex) {
            log.error("Job terminated.", ex);
            result.setFailure("Unexpected exception of type " + ex.getClass() + ": " + ex.getMessage());
        }
        return result;
    }

    // package protected for testing
    @NotNull
    JobParameters getJobParameters() {
        return new JobParametersBuilder()
            .addDate("runDate", Date.from(LocalDateTime
                .now()
                .atZone(ZoneId.systemDefault())
                .toInstant()), true)
            .toJobParameters();
    }

    private void runSingleJob(final String topic, final Job job, final SyncJobResult result,
        final JobParameters jobParameters)
        throws JobExecutionAlreadyRunningException, JobRestartException, JobInstanceAlreadyCompleteException,
        JobParametersInvalidException {
        setResultFrom(jobLauncher.run(job, jobParameters), result, topic);
    }

    private void setResultFrom(final JobExecution jobExecution, final SyncJobResult result, final String topic) {
        final Long id = jobExecution.getId();
        final String exitCode = jobExecution
            .getExitStatus()
            .getExitCode();
        final String status = jobExecution
            .getStatus()
            .name();
        final int writeCount = jobExecution
            .getStepExecutions()
            .stream()
            .mapToInt(StepExecution::getWriteCount)
            .sum();
        final String msg = String.format("Job %d has returned with exitCode %s (status %s): %d %s were synchronized.",
            id, exitCode, status, writeCount, topic);

        setMessageToResult(result, msg, jobExecution.getStatus());
    }

    private void setMessageToResult(final SyncJobResult result, final String msg, final BatchStatus status) {
        if (status == BatchStatus.COMPLETED)
            result.setSuccess(msg);
        else
            result.setFailure(msg);
    }

    private void warnAboutUnsynchronizedEntities(final SyncJobResult result) {
        String unsynchronized = warner.findUnsynchronizedPapers();
        if (unsynchronized != null)
            result.setWarning(unsynchronized);
    }
}

File: core/core-sync/src/main/kotlin/ch/difty/scipamato/core/sync/launcher/RefDataSyncJobLauncher.kt
package ch.difty.scipamato.core.sync.launcher

import ch.difty.scipamato.common.logger
import org.springframework.batch.core.BatchStatus
import org.springframework.batch.core.job.Job
import org.springframework.batch.core.job.JobExecution
import org.springframework.batch.core.job.parameters.JobParameters
import org.springframework.batch.core.job.parameters.JobParametersBuilder
import org.springframework.batch.core.launch.JobOperator
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*

private val log = logger()

/**
 * Launcher for the reference dat sync job. Consists of the following synchronization jobs:
 *
 *  1. languages
 *  1. NewStudyPage links
 *  1. code_classes
 *  1. codes
 *  1. papers
 *  1. newsletters
 *  1. newsletterTopics
 *  1. newStudyTopics
 *  1. newStudies
 *  1. keywords
 */
@Component
@Profile("!wickettest")
@Suppress("LongParameterList")
open class RefDataSyncJobLauncher(
    private val jobOperator: JobOperator,
    @Qualifier("syncLanguageJob") private val syncLanguageJob: Job,
    @Qualifier("syncNewStudyPageLinkJob") private val syncNewStudyPageLinkJob: Job,
    @Qualifier("syncCodeClassJob") private val syncCodeClassJob: Job,
    @Qualifier("syncCodeJob") private val syncCodeJob: Job,
    @Qualifier("syncPaperJob") private val syncPaperJob: Job,
    @Qualifier("syncNewsletterJob") private val syncNewsletterJob: Job,
    @Qualifier("syncNewsletterTopicJob") private val syncNewsletterTopicJob: Job,
    @Qualifier("syncNewStudyJob") private val syncNewStudyJob: Job,
    @Qualifier("syncNewStudyTopicJob") private val syncNewStudyTopicJob: Job,
    @Qualifier("syncKeywordJob") private val syncKeywordJob: Job, private val warner: Warner,
) : SyncJobLauncher {

    private var jobSuccess: BatchStatus = BatchStatus.UNKNOWN

    val jobParameters: JobParameters
        get() = JobParametersBuilder()
            .addDate("runDate", Date.from(LocalDateTime.now().atZone(ZoneId.systemDefault()).toInstant()), true)
            .toJobParameters()

    @Suppress("TooGenericExceptionCaught")
    override fun launch(
        setSuccess: (String) -> Unit,
        setFailure: (String) -> Unit,
        setWarning: (String) -> Unit,
        setDone: () -> Unit,
    ) {
        log.info { "Starting synchronization job from scipamato-core to scipamato-public..." }
        val jobParameters = jobParameters
        try {
            warner.findUnsynchronizedPapers()?.let {
                setWarning(it)
            }
            warner.findNewsletterswithUnsynchronizedPapers()?.let {
                setWarning(it)
            }
            jobOperator.start(syncLanguageJob, jobParameters).handle("languages", setSuccess, setFailure)
            jobOperator.start(syncNewStudyPageLinkJob, jobParameters).handle("newStudyPage_links", setSuccess, setFailure)
            jobOperator.start(syncCodeClassJob, jobParameters).handle("code_classes", setSuccess, setFailure)
            jobOperator.start(syncCodeJob, jobParameters).handle("codes", setSuccess, setFailure)
            jobOperator.start(syncPaperJob, jobParameters).handle("papers", setSuccess, setFailure)
            jobOperator.start(syncNewsletterJob, jobParameters).handle("newsletters", setSuccess, setFailure)
            jobOperator.start(syncNewsletterTopicJob, jobParameters).handle("newsletterTopics", setSuccess, setFailure)
            jobOperator.start(syncNewStudyTopicJob, jobParameters).handle("newStudyTopics", setSuccess, setFailure)
            jobOperator.start(syncNewStudyJob, jobParameters).handle("newStudies", setSuccess, setFailure)
            jobOperator.start(syncKeywordJob, jobParameters).handle("keywords", setSuccess, setFailure)

            log.info { "Job finished ${if (jobSuccess.isUnsuccessful) "with issues" else "successfully"}." }
            setDone()
        } catch (ex: Exception) {
            log.error(ex) { "Job terminated." }
            setFailure("Unexpected exception of type ${ex.javaClass}: ${ex.message}")
        }
    }

    private fun JobExecution.handle(
        topic: String,
        setSuccess: (String) -> Unit,
        setFailure: (String) -> Unit,
    ) {
        val msg = String.format(
            Locale.US,
            "Job $id has returned with exitCode ${exitStatus.exitCode} (" +
                "status ${this.status.name}" +
                "): ${stepExecutions.sumOf { it.writeCount }} $topic were synchronized.",
        )
        if (this.status == BatchStatus.COMPLETED) {
            setSuccess(msg)
        } else {
            setFailure(msg)
        }
    }
}
